magda-storage-api/src/createApiRouter.ts

```diff
@@ -127,23 +132,58 @@ export default function createApiRouter(options: ApiRouterOptions) {
             });
         } catch (err) {
             if (err.code === "NotFound") {
-                res.status(404).send(
-                    "No such object with fileId " +
-                        fileId +
-                        " in bucket " +
-                        bucket
-                );
+                return res
+                    .status(404)
+                    .send(
+                        "No such object with fileId " +
+                            fileId +
+                            " in bucket " +
+                            bucket
+                    );
+            } else {
+                return res.status(500).send("Unknown error");
+            }
+        }
+        const recordIdNum = res.getHeader("Record-ID"); //.toString();
+        if (recordIdNum) {
+            const recordId = recordIdNum.toString();
+            const maybeUserId = getUserId(req, options.jwtSecret);
+            let userId;
+            try {
+                userId = maybeUserId.valueOrThrow();
+            } catch (e) {
+                return res.status(400).send("No User ID set.");
+            }
+            const registryOptions: AuthorizedRegistryOptions = {
+                baseUrl: options.registryApiUrl,
+                jwtSecret: options.jwtSecret,
+                userId: userId,
+                tenantId: options.tenantId
+            };
+            const registryClient = new Registry(registryOptions);
+            const record: any = await registryClient.getRecord(
+                recordId,
+                undefined,
+                ["publishing", "dataset-access-control"]
+            );
+            // If there is an error
+            if (record.e) {
+                return res
+                    .status(404)
+                    .send(
+                        "You don't have access to this dataset/This dataset doesn't exist."
+                    );
             }
-            res.status(500).send("Unknown error");
         }
 
         const stream = await object.createStream();
         if (stream) {
             stream.on("error", _e => {
-                res.status(500).send("Unknown error");
+                return res.status(500).send("Unknown error");
             });
             stream.pipe(res);
         }
+        return undefined;
```

Comment:
Yeah we need this. Since we are explicitly `return`ing from other code paths, every possible code path needs to return something. I have made it return a `500` now.